cpp/src/cider/tests/utils/ArrowArrayBuilder.h

```diff
@@ -349,6 +349,154 @@ class ArrowArrayBuilder {
+    child_schema->name = "";
+    child_schema->format =
+        CiderBatchUtils::convertSubstraitTypeToArrowType(col_type.list().type());
+    child_schema->n_children = 0;
+    child_schema->children = nullptr;
+    child_schema->release = CiderBatchUtils::ciderEmptyArrowSchemaReleaser;
+
+    if (col_data.empty()) {
+      // append an empty buffer.
+      array_list_.push_back(nullptr);
+      schema_list_.push_back(current_schema);
+      return *this;
+    } else {
+      // check row num
+      if (row_num_ != col_data.size()) {
+        CIDER_THROW(CiderCompileException, "Row num is not equal to previous columns!");
+      }
+      CHECK_EQ(row_num_, col_data.size());
+      // check null data num
+      if (!null_data.empty()) {
+        CHECK_EQ(row_num_, null_data.size());
+      }
+      // check element num and null num for every row
+      if (!array_null_data.empty()) {
+        for (int i = 0; i < row_num_; i++) {
+          CHECK_EQ(col_data[i].size(), array_null_data[i].size());
+        }
+      }
+
+      // List array
+      current_array->length = row_num_;
+      current_array->n_children = 1;
+      current_array->offset = 0;
+      current_array->n_buffers = 2;
+      current_array->private_data = nullptr;
+      current_array->dictionary = nullptr;
+      current_array->release = CiderBatchUtils::ciderEmptyArrowArrayReleaser;
+      current_array->buffers = (const void**)allocator_->allocate(sizeof(void*) * 2);
+      current_array->children = (ArrowArray**)allocator_->allocate(sizeof(ArrowArray*));
+
+      // Validity buffer
+      size_t null_size = (row_num_ + 7) >> 3;
+      void* null_buf = (void*)allocator_->allocate(null_size);
+      std::memset(null_buf, 0xFF, null_size);
+      for (auto i = 0; i < null_data.size(); i++) {
+        if (null_data[i]) {
+          CiderBitUtils::clearBitAt((uint8_t*)null_buf, i);
+          current_array->null_count++;
+        }
+      }
+      current_array->buffers[0] = null_buf;
+
+      // Offset buffer
+      int32_t* offset_buf =
+          (int32_t*)allocator_->allocate(sizeof(int32_t) * (row_num_ + 1));
+      std::vector<int32_t> offset_vec{0};
+      for (int i = 0; i < col_data.size(); i++) {
+        if (!null_data.empty() && !null_data[i]) {
+          offset_vec.push_back(offset_vec[i] + col_data[i].size());
+        } else {
+          offset_vec.push_back(offset_vec[i]);
+        }
+      }
+      std::memcpy(offset_buf, offset_vec.data(), sizeof(int32_t) * (row_num_ + 1));
+      current_array->buffers[1] = offset_buf;
+
+      // Child values array
+      ArrowArray* child_array = new ArrowArray();
+      child_array->n_children = 0;
+      child_array->n_buffers = 2;
+      child_array->offset = 0;
+      child_array->buffers = (const void**)allocator_->allocate(sizeof(void*) * 2);
+      child_array->release = CiderBatchUtils::ciderEmptyArrowArrayReleaser;
+      child_array->private_data = nullptr;
+      child_array->dictionary = nullptr;
+
```

Comment:
seems a bit verbose in this func
- use CiderBatchUtils::allocateArrowSchema()/allocateArrowArray to reduce some initialization code？
- can call origin func to construct children arrow_array?